### PR TITLE
Cleanup registration of serialize functions and unnecessary includes

### PR DESCRIPTION
**Pull request by bvanessen**
Removed superfluous calls to #include <cereal/types/polymorphic.hpp>
in the registration of several cereal layers to reduce compilation
time.

Moving serialize functions into cereal registration cpp files to optimize compile time.

### CODE REVIEW

**Review by benson31 — COMMENTED** [review]
This is just the first batch of comments. Certainly more to come...

File: include/lbann/data_readers/sample_list_open_files_impl.hpp [hunk 1]
```diff
@@ -1,3 +1,34 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_DATA_READERS_SAMPLE_LIST_OPEN_FILES_IMPL_HPP
+#define LBANN_DATA_READERS_SAMPLE_LIST_OPEN_FILES_IMPL_HPP
+
+#include "lbann/data_readers/sample_list_open_files.hpp"
```
Comment (benson31) [inline include/lbann/data_readers/sample_list_open_files_impl.hpp:30]:
```suggestion
#include "lbann/data_readers/sample_list_open_files.hpp"
#include "lbann/data_readers/sample_list_impl.hpp" // to_sample_name_t
```

File: include/lbann/data_coordinator/io_data_buffer_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace lbann {
+
+template <typename TensorDataType>
+template <class Archive>
+void data_buffer<TensorDataType>::serialize( Archive & ar ) {
+  ar(/*CEREAL_NVP(m_input_buffers)*//*,
+                                  CEREAL_NVP(m_fetch_data_in_background),
+     CEREAL_NVP(m_data_fetch_future),
+     CEREAL_NVP(m_indices_fetched_per_mb)*/);
```
[the remark this answers is not in the document]
Comment (bvanessen) [reply include/lbann/data_coordinator/io_data_buffer_impl.hpp:40]: I believe that for now they should be commented out.

File: src/layers/cereal_registration/data_type_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+} // namespace lbann
+
+#define LBANN_INSTANTIATE_CPU_HALF
+#define LBANN_INSTANTIATE_GPU_HALF
+
+#define LBANN_CLASS_NAME data_type_layer
+#include <lbann/macros/register_template_class_with_cereal.hpp>
+
+#undef LBANN_INSTANTIATE_CPU_HALF
+#undef LBANN_INSTANTIATE_GPU_HALF
```
Comment (benson31) [inline src/layers/cereal_registration/data_type_layer.cpp:55]:
```suggestion
```

File: src/layers/learning/cereal_registration/base_convolution.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  template void base_convolution_layer<T, Device>::serialize(cereal::XMLInputArchive&);     \
+  template void base_convolution_layer<T, Device>::serialize(cereal::BinaryOutputArchive&); \
+  template void base_convolution_layer<T, Device>::serialize(cereal::BinaryInputArchive&);  \
+  template void base_convolution_layer<T, Device>::serialize(RootedXMLOutputArchive&);      \
+  template void base_convolution_layer<T, Device>::serialize(RootedXMLInputArchive&);       \
+  template void base_convolution_layer<T, Device>::serialize(RootedBinaryOutputArchive&);   \
+  template void base_convolution_layer<T, Device>::serialize(RootedBinaryInputArchive&)
+#define LBANN_INSTANTIATE_CPU_HALF
+#define LBANN_INSTANTIATE_GPU_HALF
+#include "lbann/macros/instantiate_device.hpp"
```
Comment (benson31) [inline src/layers/learning/cereal_registration/base_convolution.cpp:59]: FIXME

**Review by benson31 — COMMENTED** [review]
Next batch

**Review by benson31 — APPROVED** [review]
LGTM

**Pull request merged**